File: src/day1.pl
:- use_module(library(format)).
:- use_module(library(dcgs)).
:- use_module(library(charsio)).
:- use_module(library(pio)).
:- use_module(library(lists)).

lines([N|Ns]) --> line(N), "\n", lines(Ns).
lines([])     --> [].

line(N) --> alnums_lazy, digit(X), alnums_greedy, digit(Y), alnums_lazy, { number_chars(N, [X, Y]) }.
line(N) --> alnums_lazy, digit(X), alnums_lazy, { number_chars(N, [X, X]) }.

% If the empty variant is declared first, the grammar will be lazy/non-greedy, i.e.
% the search will continue trying to match `line` before trying other matches of `maybe_alpha`
alnums_lazy --> [].
alnums_lazy --> [X], { char_type(X, alnum) }, alnums_lazy.

% This one on the other hand is greedy
alnums_greedy --> [X], { char_type(X, alnum) }, alnums_greedy.
alnums_greedy --> [].

digit(X) --> [X], { char_type(X, decimal_digit) }.

% Part 2
digit('0') --> "zero".
digit('1') --> "one".
digit('2') --> "two".
digit('3') --> "three".
digit('4') --> "four".
digit('5') --> "five".
digit('6') --> "six".
digit('7') --> "seven".
digit('8') --> "eight".
digit('9') --> "nine".

run :-
  argv([I|_]),
  phrase_from_file(lines(Ns), I),
  portray_clause(Ns),
  sum_list(Ns, S),
  portray_clause(S),
  halt.

File: src/day1.rs
use std::io::BufRead;

const DIGIT_STRINGS: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

pub fn main(input: impl BufRead) {
    let enumerated_digit_strings = DIGIT_STRINGS.iter().zip(1..);
    let result = input
        .lines()
        .map(|line| {
            let line = line.unwrap();

            let left = enumerated_digit_strings
                .clone()
                .filter_map(|(s, i)| line.find(s).map(|j| (j, i)))
                .chain(
                    line.find(|c: char| c.is_ascii_digit())
                        .map(|j| (j, line.as_bytes()[j] - b'0')),
                )
                .min_by_key(|&(j, _)| j)
                .unwrap()
                .1 as u32;

            let right = enumerated_digit_strings
                .clone()
                .filter_map(|(s, i)| line.rfind(s).map(|j| (j, i)))
                .chain(
                    line.rfind(|c: char| c.is_ascii_digit())
                        .map(|j| (j, line.as_bytes()[j] - b'0')),
                )
                .max_by_key(|&(j, _)| j)
                .unwrap()
                .1 as u32;

            left * 10 + right
        })
        .sum::<u32>();

    println!("{result}");
}
